networking_cisco/plugins/cisco/l3/drivers/asr1k/aci_asr1k_routertype_driver.py

```diff
@@ -0,0 +1,202 @@
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from networking_cisco.plugins.cisco.extensions import routerhostingdevice
+from networking_cisco.plugins.cisco.l3.drivers.asr1k import (
+    asr1k_routertype_driver as asr1k)
+from neutron.common import constants as l3_constants
+from neutron import context as nctx
+from neutron import manager
+
+from oslo_log import log as logging
+from oslo_utils import importutils
+
+LOG = logging.getLogger(__name__)
+
+
+APIC_OWNED = 'apic_owned_'
+UUID_REGEX = '[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}'
+APIC_ML2_L3DRIVER_KLASS = (
+    'apic_ml2.neutron.services.l3_router.apic_driver.ApicL3Driver')
+GBP_L3DRIVER_KLASS = (
+    'gbpservice.neutron.services.l3_router.apic_driver.ApicGBPL3Driver')
+HOSTING_DEVICE_ATTR = routerhostingdevice.HOSTING_DEVICE_ATTR
+
+
+#ACI_ASR1K_DRIVER_OPTS = [
+#    cfg.StrOpt('aci_transit_nets_config_file', default=None,
+#               help=_("ACI with ASR transit network configuration file.")),
+#]
+#
+#cfg.CONF.register_opts(ACI_ASR1K_DRIVER_OPTS, "general")
+
+DEFAULT_EXT_DICT = {'gateway': '1.103.2.1',
+                    'host_pool_cidr': '1.103.2.0/24',
+                    'host_cidr_ver': 4}
+
+
+class AciASR1kL3RouterDriver(asr1k.ASR1kL3RouterDriver):
+
+    def __init__(self):
+        super(AciASR1kL3RouterDriver, self).__init__()
+        self._apic_driver = None
+        self._cfg_file = None
+        self._default_ext_dict = DEFAULT_EXT_DICT
+
+    @property
+    def transit_nets_cfg(self):
+        if self._cfg_file:
+            networks_dict = open(self._cfg_file, 'r').read()
+            self._transit_nets_cfg = eval(networks_dict)
+        else:
+            self._transit_nets_cfg = {}
+        return self._transit_nets_cfg
+
+    @property
+    def apic_driver(self):
+        """Get APIC driver
+
+        There are different drivers for the GBP workflow
+        and Neutron workflow for APIC. First see if the GBP
+        workflow is active, and if so get the APIC driver for it.
+        If the GBP service isn't installed, try to get the driver
+        from the Neutron (APIC ML2) workflow.
+        """
+        if not self._apic_driver:
+            try:
+                if manager.NeutronManager.get_service_plugins()[
+                    'GROUP_POLICY'].policy_driver_manager.policy_drivers[
+                        'apic'].obj:
+                    self._apic_driver = importutils.import_object(
+                        GBP_L3DRIVER_KLASS,
+                        nctx.get_admin_context_without_session()
+                    )
+                    self._apic_driver._plugin = self._l3_plugin
+            except Exception as e:
+                LOG.warn(e)
```

Comment:
The exception is there to let the user know that they've configured this router type driver, but they don't have either of the L3 plugins required in order for the driver to work. I will change this to an error.